applications/FluidDynamicsApplication/custom_elements/embedded_fluid_element_discontinuous.cpp

```diff
@@ -0,0 +1,975 @@
+{
+    rData.NumNegativeNodes = 0;
+    rData.NumPositiveNodes = NumNodes;
+    this->CalculateGeometryData(rData.PositiveSideWeights, rData.PositiveSideN, rData.PositiveSideDNDX);
+}
+
+template <class TBaseElement>
+void EmbeddedFluidElementDiscontinuous<TBaseElement>::DefineCutGeometryData(EmbeddedDiscontinuousElementData& rData) const
+{
+    // Auxiliary distance vector for the element subdivision utility
+    Vector elemental_distances = rData.ElementalDistances;
+
+    ModifiedShapeFunctions::Pointer p_calculator =
+        EmbeddedDiscontinuousInternals::GetShapeFunctionCalculator<EmbeddedDiscontinuousElementData::Dim, EmbeddedDiscontinuousElementData::NumNodes>(
+            *this, 
+            elemental_distances);
+
+    // Positive side volume
+    p_calculator->ComputePositiveSideShapeFunctionsAndGradientsValues(
+        rData.PositiveSideN, 
+        rData.PositiveSideDNDX, 
+        rData.PositiveSideWeights,
+        GeometryData::GI_GAUSS_2);
+
+    // Negative side volume
+    p_calculator->ComputeNegativeSideShapeFunctionsAndGradientsValues(
+        rData.NegativeSideN, 
+        rData.NegativeSideDNDX, 
+        rData.NegativeSideWeights,
+        GeometryData::GI_GAUSS_2);
+
+    // Positive side interface
+    p_calculator->ComputeInterfacePositiveSideShapeFunctionsAndGradientsValues(
+        rData.PositiveInterfaceN,
+        rData.PositiveInterfaceDNDX,
+        rData.PositiveInterfaceWeights,
+        GeometryData::GI_GAUSS_2);
+
+    // Negative side interface
+    p_calculator->ComputeInterfaceNegativeSideShapeFunctionsAndGradientsValues(
+        rData.NegativeInterfaceN,
+        rData.NegativeInterfaceDNDX,
+        rData.NegativeInterfaceWeights,
+        GeometryData::GI_GAUSS_2);
+
+    // Positive side interface normals
+    p_calculator->ComputePositiveSideInterfaceAreaNormals(
+        rData.PositiveInterfaceUnitNormals,
+        GeometryData::GI_GAUSS_2);
+
+    // Negative side interface normals
+    p_calculator->ComputeNegativeSideInterfaceAreaNormals(
+        rData.NegativeInterfaceUnitNormals,
+        GeometryData::GI_GAUSS_2);
+
+    // Normalize the normals
+    // Note: we calculate h here (and we don't use the value in rData.ElementSize)
+    // because rData.ElementSize might still be uninitialized: some data classes define it at the Gauss point.
+    double h = ElementSizeCalculator<Dim,NumNodes>::MinimumElementSize(this->GetGeometry());
+    const double tolerance = std::pow(1e-3 * h, Dim-1);
+    this->NormalizeInterfaceNormals(rData.PositiveInterfaceUnitNormals, tolerance);
+    this->NormalizeInterfaceNormals(rData.NegativeInterfaceUnitNormals, tolerance);
+}
+
+template <class TBaseElement>
+void EmbeddedFluidElementDiscontinuous<TBaseElement>::NormalizeInterfaceNormals(
+    typename EmbeddedDiscontinuousElementData::InterfaceNormalsType& rNormals,
+    double Tolerance) const
+{
+    for (unsigned int i = 0; i < rNormals.size(); ++i) {
+        double norm = norm_2(rNormals[i]);
```

Comment:
what happens if one is zero?

Reply:
It takes as norm the tolerance value (which is an argument of the method).
`rNormals[i] /= std::max(norm,Tolerance);`